istioctl/cmd/analyze.go

```diff
@@ -308,15 +308,25 @@ istioctl experimental analyze -L
 	return analysisCmd
 }
 
-func gatherFiles(args []string) ([]string, error) {
-	var result []string
-	for _, a := range args {
-		if _, err := os.Stat(a); err != nil {
-			return nil, fmt.Errorf("could not find file %q", a)
+func gatherFiles(args []string) ([]io.Reader, error) {
+	var readers []io.Reader
+	var r io.Reader
+	var err error
+	for _, f := range args {
+		if f == "-" {
+			if isatty.IsTerminal(os.Stdin.Fd()) {
+				return nil, fmt.Errorf("No content from stdin to read, but saw '-' arg")
+			}
+			r = os.Stdin
+		} else {
+			r, err = os.Open(f)
```

Comment:
Do we also need to close these?

In the standard CLI case I doubt it'll matter (we'll release the files when the process closes) but it's better hygiene to explicitly close. It might also matter for integ tests, which call into `cmd/analyze.go` without creating a new process. 

Maybe close each reader right after we call ReadAll on it?

Reply:
It's kind of a pain to explicitly close them since we're losing information on which Readers are files and which aren't--if we just looped over them and closed them all then we'd end up closing stdin which would be weird.

I'll add in a finalizer to make sure the files get closed at GC time, which should be sufficient unless we end up in a situation where we're leaking memory by keeping references around (at which point leaking fds isn't the biggest part of the issue)

Reply:
Sounds good. I strongly suspect closing the reader attached to stdin is harmless, but I don't know that for certain.

Reply:
If it's too much grief to get the linter happy with usage of SetFinalizer, I'm OK with not closing and just saying why we're OK with that in a comment. (Or just closing the reader on stdin as well, assuming Go does the right thing there).

Reply:
A few alternative suggestions:

1) return `io.ReadCloser` and ask the caller to close the readers
2) return a function which closes the files. The caller can defer this to ensure everything is cleaned up.